Add setup/teardown to benchmarks

GregoryTravis opened the pull request:
This change allows the addition of setup/teardown calls for individual benchmark `Spec`s:

```
        group_builder.specify "with" setup=my_setup teardown=my_teardown <|
            ...
```

The lazy data technique is sufficient when the setup requires creating data structures; the cost is amortized into the warmup period.

But sometimes it is necessary to have explicit setup and teardown for a single spec.

This PR includes an example benchmark that requires this; it will be removed before this is merged, and will be in a separate PR.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

GregoryTravis committed:
init

GregoryTravis committed:
Merge branch 'develop' into wip/gmt/12291-benchmark

GregoryTravis committed:
setup and teardown, benchmarks

GregoryTravis committed:
changelog

GregoryTravis committed:
re-enable setup/teardown

GregoryTravis committed:
setup teardown lazy

GregoryTravis committed:
api

radeusgd commented on test/Benchmarks/src/Runtime/Managed_Resources.enso:
```diff
@@ -0,0 +1,51 @@
+                Managed_Resource.register v nop_cleanup
+            mrs.map .take
+
+        with_setup =
+            some_mrs.put (make_mrs 50)
+        group_builder.specify "with" setup=with_setup teardown=clear_some_mrs <|
+            some_mrs.get.map mr-> mr.with .foo
+            Nothing
```
Looks good although the usage is very 'imperative' - requiring to rely on `Ref`.

I'd consider if we should rewrite it so that the `setup` returns a value that is then passed as argument to the bench function and teardown function (if present).

Thus bench and teardown functions would be 0-argument if setup is missing and taking 1-argument if it is present. IMO this way it'd be more 'functional'.

But given that this is a low-level library anyway I don't oppose keeping it as is, this imperative approach may also have its advantages.

GregoryTravis replied:
I think that would definitely be better, but it would require (possibly) a change to all the benchmarks, so it would be better to leave that as a separate task.

JaroslavTulach replied:
>  very 'imperative' - requiring to rely on Ref.

We certainly do not want to encourage usage of `Ref` in client code.

> bench and teardown functions would be 0-argument if setup is missing and taking 1-argument if it is present.

That's close to impossible to type properly. Thus I prefer any other solution.

> this is a low-level library

- it is true, that we don't expect our users to write benchmarks
- however the API is part of `Standard.Test` and
- if #9039 get implemented then ...
- I doubt `Standard.Test` can be called _low level library_
- quite opposite - `Standard.Test` is the most important core Enso API after `Standard.Base`
- moreover we want to have a consistency between `Bench` and `Test`
- and promoting this _imperative style_ in `Test` cases is not desirable

GregoryTravis committed:
Merge branch 'develop' into wip/gmt/12291-benchmark-setup-teardown

GregoryTravis committed:
wip

GregoryTravis committed:
Merge branch 'develop' into wip/gmt/12291-benchmark-setup-teardown

GregoryTravis committed:
add to Main

GregoryTravis committed:
Merge branch 'develop' into wip/gmt/12291-benchmark-setup-teardown

JaroslavTulach commented on distribution/lib/Standard/Test/0.0.0-dev/docs/api/Bench.md:
```diff
@@ -3,7 +3,7 @@
     - All groups:(Standard.Base.Data.Vector.Vector Standard.Test.Bench.Bench)
     - Group name:Standard.Base.Data.Text.Text configuration:Standard.Test.Bench.Bench_Options specs:(Standard.Base.Data.Vector.Vector Standard.Test.Bench.Bench)
-    - Spec name:Standard.Base.Data.Text.Text code:Standard.Base.Any.Any
+    - Spec name:Standard.Base.Data.Text.Text code:Standard.Base.Any.Any setup:(Standard.Base.Any.Any|Standard.Base.Nothing.Nothing) teardown:(Standard.Base.Any.Any|Standard.Base.Nothing.Nothing)
```
This is an incompatible change. All code that used `Spec` in `case of` statements to deconstruct the constructor needs to be updated to deconstruct additional arguments.

We don't want to prevent making incompatible changes in `Standard.Test` API in the future. What can we do to prevent such incompatible changes in the future?

JaroslavTulach replied:
- #12456 has been reported
- with the goal of 
- cleaning more than just these constructors

JaroslavTulach commented on distribution/lib/Standard/Test/0.0.0-dev/docs/api/Bench.md:
```diff
@@ -27,7 +27,7 @@
 - type Group_Builder
     - Impl name:Standard.Base.Any.Any builder:Standard.Base.Any.Any
-    - specify self name:Standard.Base.Data.Text.Text ~benchmark:Standard.Base.Any.Any -> Standard.Base.Any.Any
+    - specify self name:Standard.Base.Data.Text.Text ~benchmark:Standard.Base.Any.Any ~setup:Standard.Base.Any.Any= ~teardown:Standard.Base.Any.Any= -> Standard.Base.Any.Any
```
Why the type of these arguments is `Any`, while the type of the arguments in the constructor is different (e.g. `Any|Nothing`)?

Shouldn't the type be `Function|Nothing`? I see the type used in code is `(Any -> Any) | Nothing` - which seems OKeyish. It should be converted to `Function|Nothing`, shouldn't it?

Do we have a bug in type resolution, @Akirathan?

Akirathan replied:
[Group_Builder.specify](https://github.com/enso-org/enso/blob/08c1cd4390d349ca49d52bbf7660cfa0666908f5/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso#L72) does not specify inline type ascriptions for `setup` nor `teardown` (it is defined as `~setup=Nothing ~teardown=Nothing`), so the *inferred* type is `Any`.

[Bench.Spec constructor](https://github.com/enso-org/enso/blob/08c1cd4390d349ca49d52bbf7660cfa0666908f5/distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso#L85) has inline type ascriptions for the `setup` and `teardown` fields.

From the perspective of *API signature check*, everything is fine - the type ascriptions are just written to the markdown files. From the perspective of *Static type checking*, it is probably fine as well. But we don't know that for sure since the static type checking is not run on the CI yet.

I don't understand what kind of bug are you seeing? Could you explain more please?

JaroslavTulach replied:
The [signature of setup](https://github.com/enso-org/enso/pull/12424/files#diff-882c05fa74bd3245ba9026ea09a161b16da90e144eaaad590c911204a86c9d62R85) in code is `(setup : (Any -> Any) | Nothing)`:
- that shouldn't be `Any|Nothing` as the signature files currently contain
- it **should be** `Function|Nothing`

Shouldn't it? E.g. we don't handle arity and argument types, but we should convert any function like definition to `Function`, shouldn't we? I do remember doing that somewhere, but apparently it doesn't happen in the type signatures.

Akirathan replied:
You are right. Created https://github.com/enso-org/enso/issues/12453 for that.

JaroslavTulach commented on distribution/lib/Standard/Test/0.0.0-dev/src/Bench.enso:
```diff
@@ -122,7 +122,7 @@ type Bench
                 conf.validate
                 ensure_distinct_names <| specs.map (_.name)
-            Bench.Spec name code ->
+            Bench.Spec name code _ _ ->
```
This change is only necessary because _adding additional fields_ to non-private constructors is an **incompatible change**.

JaroslavTulach commented on test/Benchmarks/src/Runtime/Managed_Resources.enso:
```diff
@@ -0,0 +1,51 @@
+
+collect_benches = Bench.build builder->
+    builder.group "Managed_Resource" options group_builder->
+        data = Data.create
```
The proper way to deal with _lazy setup_ for `group` is to define own `type` like `Data` with lazy fields and use them when needed. Let's built upon that and:
- define another type `DataForBench` with lazy field
- make `some_mrs` a lazy field in `DataForBench`
- no need for `setup` then - remove it
- in `teardown` `clear_some_mrs`

JaroslavTulach replied:
- there already seems to be `teardown` in #9035
- inside of `Group`: https://github.com/enso-org/enso/blob/aa2e77e06f6bec2be003c3437340a8725d736589/distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso#L43
- the `Bench` teardown shall be **consistent** with the `Test` approach
- btw., @Akirathan, how can `Group` be `private` (not listed in the `docs/api` at all) and its `teardown` still be available (as #9035 suggests)?

Akirathan replied:
> btw., how can Group be private (not listed in the docs/api at all) and its teardown still be available (as https://github.com/enso-org/enso/issues/9035 suggests)?

Because [Suite_Builder.group](https://github.com/enso-org/enso/blob/48c17845a78f62ff244cb332d3571790b14571fb/distribution/lib/Standard/Test/0.0.0-dev/src/Suite.enso#L37) exposes `Group_Builder` to the caller, so the caller can call `Group_Builder.teardown` even if it is private.

JaroslavTulach reviewed (commented):
### Avoid non-functional coding patterns

- given the only use of `setup` is via `Ref`
- which isn't a code style we want to promote
- I **request** (convince me if you don't agree): let's **not introduce** `setup` and add only `teardown`
- use lazy constructor fields (as we did so far) to delay _setup_
- we don't have `setup` for `Group` either...

### Keep consistency of `Bench` and `Test`

- both `Test` and `Bench` model the same hierarchical structure
- both need to solve the same issues
- whatever solution we come with we should add it to `Test` as well
- **opinion** (ignore if you want)

### Avoid Incompatible Changes

- adding fields to public constructor is an **incompatible change**
- we shouldn't be making incompatible changes in core Enso APIs
- definitely when #9039 is implemented
- get ready for that (by _"last incompatible change"_ - e.g. those famous words) and _make the constructors_ of `Bench` `private`
- **advice** (if you don't agree, please explain why)

Akirathan reviewed (changes requested):
@JaroslavTulach does not like the incompatible change in the Enso API this PR adds, but I have even bigger concern. Your PR only changes how the [Running stdlib benchmarks from Enso](https://github.com/enso-org/enso/blob/701bba6504e315b34e3ac8454b794002cccada8d/docs/infrastructure/benchmarks.md#L104) behave, but not how [Running them via JMH](https://github.com/enso-org/enso/blob/701bba6504e315b34e3ac8454b794002cccada8d/docs/infrastructure/benchmarks.md#L128) behaves, which is much more important since that is what CI does.

If you run, e.g., `sbt:std-benchmarks> run -l` and look into the Java source that was generated in `std-bits/benchmarks/target/scala-2.13/src_managed`, (for convenience, I have uploaded one of these generated files as [gist](https://gist.github.com/Akirathan/3c54c2b29ce479d7c749ab65593e010d#file-collections-java)), you will see that these sources already have their own `setup` and `teardown`. And they **do not call** the setup and teardown that you have introduced.

These Java JMH sources are generated by [BenchProcessor.java](https://github.com/enso-org/enso/blob/9182f91e35ad33ec7ffe91ce22880218dc2effc3/lib/scala/bench-processor/src/main/java/org/enso/benchmarks/processor/BenchProcessor.java#L192), so if you would really like to introduce custom setup and teardown, you would have to modify [SpecCollector.generateClassForGroup](https://github.com/enso-org/enso/blob/0e5406227a1b737821484e5100eb266495d84c64/lib/scala/bench-processor/src/main/java/org/enso/benchmarks/processor/SpecCollector.java#L232).

If you insist that the setup and teardown as you suggest is introduced, we would have to change `SpecCollector` as well. And looking at your [Managed_Resources benchmark](https://github.com/enso-org/enso/blob/08c1cd4390d349ca49d52bbf7660cfa0666908f5/test/Benchmarks/src/Runtime/Managed_Resources.enso) I don't see why it cannot be implemented with the *lazy data creation* mechanism that we already have.

GregoryTravis committed:
Merge branch 'develop' into wip/gmt/12291-benchmark-setup-teardown

GregoryTravis commented:
> ### Avoid non-functional coding patterns
> * given the only use of `setup` is via `Ref`
> * which isn't a code style we want to promote
> * I **request** (convince me if you don't agree): let's **not introduce** `setup` and add only `teardown`
> * use lazy constructor fields (as we did so far) to delay _setup_
> * we don't have `setup` for `Group` either...

I agree that using `Ref` is bad; Radek's suggestion that setup could pass data somehow to the benchmark is better, but that is a bigger change we don't want to do now.

Indeed I first wrote this benchmark using lazy data. But I realized I needed a `teardown`, so that we aren't leaving a bunch of registered managed resources around that could distort other benchmarks.

I don't 100% agree that lazy data is ideal for this, because we are not just creating data, we are inducing a particular state within the engine. If someone were to add another benchmark here, and re-use the MRs created in the existing benchmark, because it's created by lazy data, they won't be created, and the benchmark will fail with an uninitialized state. Each lazy data field can be used only once -- this is not how lazily-created data works.

Each benchmark will have to have its own lazy data field, which is fine, since this is an unusual case -- I'll add comments to make this clear.

So what I will do now:
* Remove `setup`
* Use lazy data fields for setup
* No more use of `Ref`
* Keep `Group.teardown`, similar to what we have in `Test`
* Make `Bench` constructors private

I think this addresses all the concerns.

GregoryTravis commented:
> So what I will do now:

Also:
* Implement teardown in JMH benchmarks as well

GregoryTravis commented:
After discussion with @Akirathan, I now understand that `teardown` isn't really needed for this benchmark. Under JMH, every individual benchmark is run with its own polyglot `Context`, so cleanup is unnecessary.

This *won't* apply to running benchmarks locally via the Enso benchmark runner. Since local benchmarks are not definitive, and are usually run individually, cleanup is unecessary, so I can just remove it.

What do you think, @JaroslavTulach @radeusgd ?

GregoryTravis commented:
Closing as the teardown is not necessary for JMH benchmarks.
For Enso benchmarks, the `Managed_Resource` benchmark will leave registered resources around for the next benchmark, until GC cleans them up, but since locally-run benchmarks are not definitive and because they are almost always run individually, this is also not a concern.
